File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
var lengthOfLongestSubstring = function(s) {
    let i = 0
    let j = 0
    let output = 0
    let lookup = {}

    while (j < s.length) {

        while (s[j] in lookup) {
            if (lookup[s[i]] == 1) delete lookup[s[i]]
            else lookup[s[i]] -= 1
            i += 1
        }

        if (!(s[j] in lookup)) lookup[s[j]] = 0
        lookup[s[j]] += 1
        output = Math.max(output, j - i + 1)
        j += 1
    }   
    return output
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # output = 0
        # def isUnique(arr):
        #     lookup = set(arr)

        #     return len(arr) == len(lookup)

        # for i in range(len(s)):
        #     for j in range(i, len(s)):
        #         substr = s[i: j + 1]
        #         if (isUnique([char for char in substr])):
        #             output = max(output, len(substr))
        
        # return output


        lookup = set()
        i = 0
        j = 0
        output = 0

        while (j < len(s)):

            while (s[j] in lookup):
                lookup.remove(s[i])
                i += 1

            lookup.add(s[j])
            length = j - i + 1
            output = max(output, length)
            j += 1

        return output

        